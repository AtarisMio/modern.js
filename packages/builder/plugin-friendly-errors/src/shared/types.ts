import StackTracey from 'stacktracey';
import type * as webpack from 'webpack';

export type TracingFrame = StackTracey.Entry;

export interface ParsedError extends webpack.WebpackError {
  raw: webpack.WebpackError;
  trace: TracingFrame[];
}

export type ErrorTransformer = (error: ParsedError) => ParsedError | void;

export type ErrorFormatter = (error: ParsedError) => string | void;
