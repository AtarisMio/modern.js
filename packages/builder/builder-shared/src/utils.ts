<<<<<<< HEAD
import _ from '@modern-js/utils/lodash';

export function deepFreezed<T>(obj: T): T {
  if (!_.isObject(obj) || _.isNull(obj)) {
    return obj;
  }
  if (Array.isArray(obj)) {
    obj.forEach(deepFreezed);
  } else {
    Object.values(obj).forEach(deepFreezed);
  }
  return Object.freeze(obj);
}

export const extendsType =
  <T>() =>
  <P extends T>(source: P): P =>
    source;
=======
import assert from 'assert';
import _ from '@modern-js/utils/lodash';
import { CreateBuilderOptions } from './types/builder';

export function pick<T, U extends keyof T>(obj: T, keys: ReadonlyArray<U>) {
  return keys.reduce((ret, key) => {
    if (obj[key] !== undefined) {
      ret[key] = obj[key];
    }
    return ret;
  }, {} as Pick<T, U>);
}

export function deepFreezed<T extends Record<any, any> | any[]>(obj: T): T {
  assert(typeof obj === 'object');
  const handle = (item: any) =>
    typeof item === 'object' ? deepFreezed(item) : item;
  const ret = (
    Array.isArray(obj) ? _.map(obj, handle) : _.mapValues(obj, handle)
  ) as T;
  return Object.freeze(ret);
}

export function applyDefaultBuilderOptions(options?: CreateBuilderOptions) {
  return {
    cwd: process.cwd(),
    entry: {},
    target: ['web'],
    configPath: null,
    framework: 'modern-js',
    ...options,
  } as Required<CreateBuilderOptions>;
}
>>>>>>> ac5486156 (refactor(builder): split builder and provider (#1804))
