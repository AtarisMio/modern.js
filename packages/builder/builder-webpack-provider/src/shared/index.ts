export * from './fs';
export * from './utils';
export * from './format';
<<<<<<< HEAD
=======
export * from './constants';
export * from './browserslist';
>>>>>>> ac5486156 (refactor(builder): split builder and provider (#1804))
