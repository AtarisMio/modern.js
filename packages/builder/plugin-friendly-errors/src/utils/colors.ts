import chalk from 'chalk';
import { Severity } from '../formatters/defaultError';

export function formatTitle(severity: Severity, message: string) {
  return chalk[bgColor(severity)].black('', message, '');
}

export function formatText(severity: Severity, message: string) {
  return chalk[textColor(severity)](message);
}

export function bgColor(
  severity: Severity,
): `bg${Capitalize<ReturnType<typeof textColor>>}` {
  const color = textColor(severity);
  const colorMark = capitalizeFirstLetter(color);
  return `bg${colorMark}` as any;
}

export function textColor(severity: Severity) {
  switch (severity.toLowerCase()) {
    case 'success':
      return 'green';
    case 'info':
      return 'blue';
    case 'note':
      return 'white';
    case 'warning':
      return 'yellow';
    case 'error':
      return 'red';
    default:
      return 'red';
  }
}

export function capitalizeFirstLetter(str: string) {
  return str.charAt(0).toUpperCase() + str.slice(1);
}
