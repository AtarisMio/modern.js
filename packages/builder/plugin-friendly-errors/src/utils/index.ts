/**
 * Concat and flattens non-null values.
 * Ex: concat(1, undefined, 2, [3, 4]) = [1, 2, 3, 4]
 */
export function concat<T>(...args: T[]): T[] {
  const _args = Array.from(args).filter(e => e != null);
  const baseArray = Array.isArray(_args[0]) ? _args[0] : [_args[0]];
  return Array.prototype.concat.apply(baseArray, _args.slice(1));
}

/**
 * Dedupes array based on criterion returned from iteratee function.
 * Ex: uniqueBy(
 *     [{ id: 1 }, { id: 1 }, { id: 2 }],
 *     val => val.id
 * ) = [{ id: 1 }, { id: 2 }]
 */
export function uniqueBy<T>(arr: T[], func: (val: T) => any) {
  const seen: Record<any, number> = {};
  return arr.filter(el => {
    const e = func(el);
    return !(e in seen) && (seen[e] = 1);
  });
}
