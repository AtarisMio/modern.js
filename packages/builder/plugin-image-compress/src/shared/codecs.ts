import {
  compressJpeg,
  losslessCompressPng,
  pngQuantize,
  Transformer,
} from '@napi-rs/image';
import { Codec, Codecs } from '../types';

export const jpegCodec: Codec<'jpeg'> = {
  handler(buf, options) {
    return compressJpeg(buf, options);
  },
  defaultOptions: {
    test: /\.jpeg$/,
  },
};

export const pngCodec: Codec<'png'> = {
  handler(buf, options) {
    return pngQuantize(buf, options);
  },
  defaultOptions: {
    test: /\.png$/,
  },
};

export const pngLosslessCodec: Codec<'pngLossless'> = {
  handler(buf, options) {
    return losslessCompressPng(buf, options);
  },
  defaultOptions: {
    test: /\.png$/,
  },
};

export const webpCodec: Codec<'webp'> = {
  handler(buf, options) {
    return new Transformer(buf).webp(options.quality);
  },
  defaultOptions: {
    test: /\.webp$/,
  },
};

export const webpLosslessCodec: Codec<'webpLossless'> = {
  handler(buf) {
    return new Transformer(buf).webpLossless();
  },
  defaultOptions: {
    test: /\.webp$/,
  },
};

export const avifCodec: Codec<'avif'> = {
  handler(buf, options) {
    return new Transformer(buf).avif(options);
  },
  defaultOptions: {
    test: /\.avif$/,
  },
};

export const icoCodec: Codec<'ico'> = {
  handler(buf) {
    return new Transformer(buf).ico();
  },
  defaultOptions: {
    test: /\.(ico|icon)$/,
  },
};

const codecs: Record<Codecs, Codec<any>> = {
  jpeg: jpegCodec,
  png: pngCodec,
  pngLossless: pngLosslessCodec,
  webp: webpCodec,
  webpLossless: webpLosslessCodec,
  avif: avifCodec,
  ico: icoCodec,
};

export default codecs;
