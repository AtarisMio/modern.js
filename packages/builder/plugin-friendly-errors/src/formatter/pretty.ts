import chalk from '@modern-js/utils/chalk';
import { ErrorFormatter } from '../shared/types';

export const prettyFormatter: ErrorFormatter = e => {
  const message = e.message || e;
  const name = e.name || 'Error';

  const formattedStack = e.trace
    .map(frame => {
      const { file, line, column } = frame;
      return `  at ${chalk.gray(file)}:${line}:${column}`;
    })
    .join('\n');

  return `${chalk.red(`${name}: ${message}`)}\n\n${formattedStack}`;
};
