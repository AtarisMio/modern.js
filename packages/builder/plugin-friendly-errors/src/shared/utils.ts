import assert from 'assert';
import { WebpackError } from 'webpack';
import StackTracey from 'stacktracey';
import { baseFormatter } from '../formatter/base';
import { ErrorFormatter, ErrorTransformer, ParsedError } from './types';

const _traceSerializeCache = new WeakMap<WebpackError, ParsedError>();

export const parseError = (error: WebpackError) => {
  if (_traceSerializeCache.has(error)) {
    return _traceSerializeCache.get(error)!;
  }

  const parsed = cloneObject(error) as ParsedError;
  parsed.trace = new StackTracey(error).items;
  parsed.raw = error;

  _traceSerializeCache.set(error, parsed);

  return parsed;
};

export const cloneObject = <T>(error: T): T => {
  const cloned = Object.create(Object.getPrototypeOf(error));
  Object.assign(cloned, error);
  return cloned;
};

export const transformError = (
  transformers: ErrorTransformer[],
  error: ParsedError,
): WebpackError => {
  return transformers.reduce(
    (error, transformer) => transformer(error) || error,
    error,
  );
};

export const formatError = (
  formatters: ErrorFormatter[] = [baseFormatter],
  error: ParsedError,
) => {
  let formatted: string | void;
  for (const formatter of formatters) {
    formatted = formatter(error);
    if (formatted) {
      break;
    }
  }
  assert(typeof formatted === 'string', 'No error formatter found');
  return formatted;
};
