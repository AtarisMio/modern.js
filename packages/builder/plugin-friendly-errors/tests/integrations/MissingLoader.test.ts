import assert from 'assert';
import ModuleParseError from 'webpack/lib/ModuleParseError';
import { createStubBuilder } from '@modern-js/builder-webpack-provider/stub';
import { useFixture } from '@modern-js/e2e';
import { test, expect } from 'vitest';
import { baseFormatter } from '@/formatter';
import { parseError } from '@/shared/utils';

test('MissingLoader', async () => {
  const options = await useFixture('@modern-js/e2e/fixtures/builder/basic');
  const builder = await createStubBuilder({
    ...options,
    plugins: 'minimal',
  });
  const [{ stats }] = await builder.unwrapHook('onAfterBuildHook');
  assert(stats && 'compilation' in stats);
  const source = [
    'import { FC } from "react";',
    'export const Foo: FC = () => <div>foo</div>;',
  ].join('\n');
  const rawError = new Error('foo');
  const error = new ModuleParseError(
    source,
    rawError,
    ['bar-loader'],
    'javascript/esm',
  );
  const parsed = parseError(error);
  expect(baseFormatter(parsed)).toMatchSnapshot();
});
