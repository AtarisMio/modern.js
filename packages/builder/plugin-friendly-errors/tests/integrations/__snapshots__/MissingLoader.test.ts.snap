// Vitest Snapshot v1

exports[`MissingLoader 1`] = `
"/moduleparseerror Module parse failed: foo
File was processed with these loaders:
 * bar-loader
You may need an additional loader to handle the result of these loaders.
Error: foo
    at /Users/bytedance/repositories/modern.js/packages/builder/plugin-friendly-errors/tests/integrations/MissingLoader.test.ts:27:20
    at async runTest (/Users/bytedance/repositories/modern.js/node_modules/<PNPM_INNER>/vitest/dist/chunk-runtime-error.87a2b5a2.mjs:488:5)
    at async runSuite (/Users/bytedance/repositories/modern.js/node_modules/.pnpm/vitest@0.21.1_@vitest+ui@0.21.1/node_modules/vitest/dist/chunk-runtime-error.87a2b5a2.mjs:576:13)
    at async runFiles (/Users/bytedance/repositories/modern.js/node_modules/.pnpm/vitest@0.21.1_@vitest+ui@0.21.1/node_modules/vitest/dist/chunk-runtime-error.87a2b5a2.mjs:620:5)
    at async startTestsNode (/Users/bytedance/repositories/modern.js/node_modules/.pnpm/vitest@0.21.1_@vitest+ui@0.21.1/node_modules/vitest/dist/chunk-runtime-error.87a2b5a2.mjs:638:3)
    at async /Users/bytedance/repositories/modern.js/node_modules/.pnpm/vitest@0.21.1_@vitest+ui@0.21.1/node_modules/vitest/dist/entry.mjs:76:9
    at async Module.withEnv (/Users/bytedance/repositories/modern.js/node_modules/.pnpm/vitest@0.21.1_@vitest+ui@0.21.1/node_modules/vitest/dist/chunk-runtime-error.87a2b5a2.mjs:259:5)
    at async run (/Users/bytedance/repositories/modern.js/node_modules/.pnpm/vitest@0.21.1_@vitest+ui@0.21.1/node_modules/vitest/dist/entry.mjs:71:5)
    at async file:/Users/bytedance/repositories/modern.js/node_modules/.pnpm/tinypool@0.2.4/node_modules/tinypool/dist/esm/worker.js:99:20
    at /Users/bytedance/repositories/modern.js/packages/builder/plugin-friendly-errors/tests/integrations/MissingLoader.test.ts:28:17
    at async runTest (/Users/bytedance/repositories/modern.js/node_modules/.pnpm/vitest@0.21.1_@vitest+ui@0.21.1/node_modules/vitest/dist/chunk-runtime-error.87a2b5a2.mjs:488:5)
    at async runSuite (/Users/bytedance/repositories/modern.js/node_modules/.pnpm/vitest@0.21.1_@vitest+ui@0.21.1/node_modules/vitest/dist/chunk-runtime-error.87a2b5a2.mjs:576:13)
    at async runFiles (/Users/bytedance/repositories/modern.js/node_modules/.pnpm/vitest@0.21.1_@vitest+ui@0.21.1/node_modules/vitest/dist/chunk-runtime-error.87a2b5a2.mjs:620:5)
    at async startTestsNode (/Users/bytedance/repositories/modern.js/node_modules/.pnpm/vitest@0.21.1_@vitest+ui@0.21.1/node_modules/vitest/dist/chunk-runtime-error.87a2b5a2.mjs:638:3)
    at async /Users/bytedance/repositories/modern.js/node_modules/.pnpm/vitest@0.21.1_@vitest+ui@0.21.1/node_modules/vitest/dist/entry.mjs:76:9
    at async Module.withEnv (/Users/bytedance/repositories/modern.js/node_modules/.pnpm/vitest@0.21.1_@vitest+ui@0.21.1/node_modules/vitest/dist/chunk-runtime-error.87a2b5a2.mjs:259:5)
    at async run (/Users/bytedance/repositories/modern.js/node_modules/.pnpm/vitest@0.21.1_@vitest+ui@0.21.1/node_modules/vitest/dist/entry.mjs:71:5)
    at async file:/Users/bytedance/repositories/modern.js/node_modules/.pnpm/tinypool@0.2.4/node_modules/tinypool/dist/esm/worker.js:99:20"
`;
